### DISCUSSION TITLE
Subscribing to a ProSys OPCUA server node id using the namespace uri

### DISCUSSION THREAD

**Discussion by KishanKishore** [FreeOpcUa/opcua-asyncio#588] [Q&A]
I am trying to connect to a ProSys OPCUA server's node using the nodeId: ```nsu=http://www.prosysopc.com/OPCUA/SimulationNodes/;i=1001```. When executing the following code:
```
from asyncua import Client
client = Client("opc.tcp://localhost:53530/OPCUA/SimulationServer")
await client.connect()
node = client.get_node("nsu=http://www.prosysopc.com/OPCUA/SimulationNodes/;i=1001")
print(await node.get_access_level())
```

I get the following error:
```
Received an error: MessageAbort(error:StatusCode(BadDecodingError), reason:Bad_DecodingError (code=0x80070000, description="Unexpected NodeId Encoding Byte -126"))
Received an error: MessageAbort(error:StatusCode(BadDecodingError), reason:Bad_DecodingError (code=0x80070000, description="Unexpected NodeId Encoding Byte -126"))
Exception raised while parsing message from server
Traceback (most recent call last):
  File "/Users/kishan/Documents/Configuration/edge-sdk/lib/python3.7/site-packages/asyncua/client/ua_client.py", line 169, in _call_callback
    self._callbackmap[request_id].set_result(body)
KeyError: 0

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/kishan/Documents/Configuration/edge-sdk/lib/python3.7/site-packages/asyncua/client/ua_client.py", line 81, in _process_received_data
    self._process_received_message(msg)
  File "/Users/kishan/Documents/Configuration/edge-sdk/lib/python3.7/site-packages/asyncua/client/ua_client.py", line 105, in _process_received_message
    self._call_callback(0, ua.UaStatusCodeError(msg.Error.value))
  File "/Users/kishan/Documents/Configuration/edge-sdk/lib/python3.7/site-packages/asyncua/client/ua_client.py", line 171, in _call_callback
    raise ua.UaError(f"No request found for request id: {request_id}, pending are {self._callbackmap.keys()}")
asyncua.ua.uaerrors._base.UaError: No request found for request id: 0, pending are dict_keys([4])
---------------------------------------------------------------------------
ConnectionError                           Traceback (most recent call last)
<ipython-input-25-d3da8efff58d> in async-def-wrapper()

~/Documents/Configuration/edge-sdk/lib/python3.7/site-packages/asyncua/common/node.py in get_access_level(self)
    113         Get the access level attribute of the node as a set of AccessLevel enum values.
    114         """
--> 115         result = await self.read_attribute(ua.AttributeIds.AccessLevel)
    116         return ua.AccessLevel.parse_bitfield(result.Value.Value)
    117

~/Documents/Configuration/edge-sdk/lib/python3.7/site-packages/asyncua/common/node.py in read_attribute(self, attr)
    298         params = ua.ReadParameters()
    299         params.NodesToRead.append(rv)
--> 300         result = await self.server.read(params)
    301         result[0].StatusCode.check()
    302         return result[0]

~/Documents/Configuration/edge-sdk/lib/python3.7/site-packages/asyncua/client/ua_client.py in read(self, parameters)
    359         request = ua.ReadRequest()
    360         request.Parameters = parameters
--> 361         data = await self.protocol.send_request(request)
    362         response = struct_from_binary(ua.ReadResponse, data)
    363         self.logger.debug(response)

~/Documents/Configuration/edge-sdk/lib/python3.7/site-packages/asyncua/client/ua_client.py in send_request(self, request, timeout, message_type)
    148         except Exception:
    149             if self.state != self.OPEN:
--> 150                 raise ConnectionError("Connection is closed") from None
    151
    152             raise

ConnectionError: Connection is closed
```

What am I doing incorrectly here? If I try the same with ```ns=3;i=1001```, it works as expected.

Here are some screenshots from my local OPCUA server/client for reference: 

<img width="1062" alt="Screenshot 2021-06-03 at 3 44 01 PM" src="https://user-images.githubusercontent.com/59863682/120626632-920c3c00-c482-11eb-9253-2ad396dc2670.png">
<img width="1122" alt="Screenshot 2021-06-03 at 3 44 28 PM" src="https://user-images.githubusercontent.com/59863682/120626648-97698680-c482-11eb-8fb8-8ae7990eb6f8.png">

**Comment by AndreasHeine** [comment]
client.get_node(ua.NodeId.from_string(...))

https://github.com/FreeOpcUa/opcua-asyncio/blob/7d7841bfb7b4e351797b8a5cebdfa68a6418e406/asyncua/ua/uatypes.py#L464

should return a expanded node id?

https://github.com/FreeOpcUa/opcua-asyncio/blob/7d7841bfb7b4e351797b8a5cebdfa68a6418e406/asyncua/ua/uatypes.py#L584

**Reply by KishanKishore** [reply]
I tried doing this. But I still get the same error. 
```
client = Client("opc.tcp://localhost:53530/OPCUA/SimulationServer")
await client.connect()
node_1 = ua.NodeId.from_string("nsu=http://www.prosysopc.com/OPCUA/SimulationNodes/;i=1001")
node_2 = client.get_node(node_1)
print(await node_2.get_access_level())
```

Moreover if I understand correctly that method is already called in the constructor of the ```Node``` class here: 
https://github.com/FreeOpcUa/opcua-asyncio/blob/7d7841bfb7b4e351797b8a5cebdfa68a6418e406/asyncua/common/node.py#L55

**Reply by AndreasHeine** [reply]
hm 0x80070000 -> looks like server complains he cant decode the nodeid

**Reply by KishanKishore** [reply]
So, could it be a problem with the ProSys server ?

**Reply by AndreasHeine** [reply]
not sure have not used it since it was changed to dataclasses!
would bet its on our side...

**Reply by AndreasHeine** [reply]
can you try it with v0.9.10 or older?

**Reply by KishanKishore** [reply]
Same result. I will also try this on some other ProSys OPCUA server just to confirm if it is a problem with a specific server.

**Reply by AndreasHeine** [reply]
hm then its not related to dataclasses... and its again open ^^
is it the prosys simulation server? version? java stackversion?

**Reply by KishanKishore** [reply]
I am using the ProSys Simulation Server (on MacOS) with the following version: 
<img width="523" alt="Screenshot 2021-06-03 at 6 56 55 PM" src="https://user-images.githubusercontent.com/59863682/120650551-9f82ef80-c49d-11eb-994e-d9a0be320451.png">

Here information related to Java: 
```
openjdk version "1.8.0_242"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_242-b08)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.242-b08, mixed mode)
```

**Reply by AndreasHeine** [reply]
its a bug on our side...

**Reply by AndreasHeine** [reply]
in 0.9.14 we dont differenciate between NodeId class and ExpandedNodeId class 

Latest:
https://github.com/FreeOpcUa/opcua-asyncio/blob/7a86b031f2ae23463dccaaf2ca10a4b74a044e71/asyncua/ua/uatypes.py#L464

Master:
https://github.com/FreeOpcUa/opcua-asyncio/blob/7d7841bfb7b4e351797b8a5cebdfa68a6418e406/asyncua/ua/uatypes.py#L584

**Reply by KishanKishore** [reply]
Should I raise an issue in the issues section?

**Reply by AndreasHeine** [reply]
@KishanKishore working on a fix: https://github.com/FreeOpcUa/opcua-asyncio/pull/589

**Reply by AndreasHeine** [reply]
the decode error happens because the field order is different in NodeId and ExpandedNodeId now i made the order right an some other errors occure... need to fix them as well... hang on ;)

**Reply by KishanKishore** [reply]
Thanks @AndreasHeine for the help!

**Comment by oroulet** [comment]
@KishanKishore My current conclusion is that you cannot use an ExpandedNode in a ReadRequest. see the last comments in https://github.com/FreeOpcUa/opcua-asyncio/pull/589 . You are welcome to prove me otherwise. Maybe try another client implementation and make ReadRequest that works with an ExpandedNodeId

**Reply by KishanKishore** [reply]
@oroulet Thanks for the reply. Has this been mentioned somewhere in the specification ? I went through some articles regarding the different types of node id representation but none of them mentioned about this restriction. I will go through the comments in the PR as well.

I will try to use a different implementation as well.

**Reply by oroulet** [reply]
I am not 100% sure but i notice that some structures explicitly support expandedNodeId so i understand that the others do not.
